### PR TITLE
Mark branin test as slow running and add simpler quadratic test

### LINKED ISSUE #256: Run additional tests on develop after merging
As trieste grows, and as we improve our currently limited integration testing, it will no longer be feasible to run all the tests before every commit. Relatedly, we don't want to dissuade people from writing new tests for fear they would slow down development.

One solution is to have a suite of tests that are only run after merging.

* Consistently reproducible tests that take no more than a few hours can be run automatically after each merge to develop, allowing us to link any regressions to the offending commit.
* Inconsistent or very long running tests can be run on a schedule, allowing us to observe gradual regressions or intermittent issues.

### PR DESCRIPTION
**Pull request by uri-granta**
when reviewing this make sure to add ?w=1 to the url to ignore whitespace

### CODE REVIEW

File: tests/integration/test_bayesian_optimization.py [hunk 1]
```diff
@@ -142,58 +140,101 @@
...
+                    )
+                    # these accuracies are the current best for the given number of optimization
+                    # steps, which makes this is a regression test
+                    assert tf.reduce_any(
+                        tf.reduce_all(relative_minimizer_err < 0.05, axis=-1), axis=0
+                    )
+                    npt.assert_allclose(best_y, SCALED_BRANIN_MINIMUM, rtol=0.005)
+                else:
+                    # currently assume that every rule should be able to solve this in 5 steps
+                    npt.assert_allclose(best_x, tf.constant([1.0, 1.0], tf.float64), rtol=0.05)
```
Comment (hstojic) [inline tests/integration/test_bayesian_optimization.py]: we might need to pass tolerances as arguments, if we want to have different objectives, no?

Comment (uri-granta) [reply tests/integration/test_bayesian_optimization.py]: There are only two hard-coded objectives (controlled by the branin parameter), each with its own hard-coded tolerance. If we need more flexibility we can easily refactor.

**Review by hstojic — APPROVED** [review]
looks good so far...
is there a specific reason for defining these tests in a class?

**Comment by uri-granta** [comment]
> looks good so far... is there a specific reason for defining these tests in a class?

I originally tried to parametrize the class directly (which should be possible), but that didn't work. Personally I quite like grouping related tests into classes, but not too fussed either way.

**Comment by hstojic** [comment]
> > looks good so far... is there a specific reason for defining these tests in a class?
> 
> I originally tried to parametrize the class directly (which should be possible), but that didn't work. Personally I quite like grouping related tests into classes, but not too fussed either way.

Since we aren't doing that with any other test, for consistency I wouldn't use a class here if its not necessary

File: tests/integration/test_ask_tell_optimization.py [hunk 2]
```diff
@@ -46,9 +46,7 @@
 from trieste.space import Box, SearchSpace
 from trieste.types import State, TensorType
 
-
-@random_seed
-@pytest.mark.parametrize(
+BRANIN_OPTIMIZER_PARAMS = (
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]: How about more neutral name, e.g. `TEST_SCENARIOS` or simply `OPTIMIZER_PARAMS`? They are not super specific to branin, and we use them as inputs to both branin and quadratic tests

File: tests/integration/test_ask_tell_optimization.py [hunk 3]
```diff
@@ -97,7 +95,39 @@
...
+    acquisition_rule_fn: Callable[[], AcquisitionRule[TensorType, SearchSpace]]
+    | Callable[
+        [],
+        AcquisitionRule[State[TensorType, AsynchronousRuleState | TrustRegion.State], Box],
+    ],
+) -> None:
+    _test_ask_tell_optimization_finds(False, min(num_steps, 5), reload_state, acquisition_rule_fn)
+
+
+def _test_ask_tell_optimization_finds(
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]:
```suggestion
def _test_ask_tell_optimization_finds_minima(
```

File: tests/integration/test_ask_tell_optimization.py [hunk 4]
```diff
@@ -97,7 +95,39 @@
...
+def test_ask_tell_optimizer_finds_minima_of_simple_quadratic(
+    num_steps: int,
+    reload_state: bool,
+    acquisition_rule_fn: Callable[[], AcquisitionRule[TensorType, SearchSpace]]
+    | Callable[
+        [],
+        AcquisitionRule[State[TensorType, AsynchronousRuleState | TrustRegion.State], Box],
+    ],
+) -> None:
+    _test_ask_tell_optimization_finds(False, min(num_steps, 5), reload_state, acquisition_rule_fn)
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]: Move this 5 to a constant, e.g. `simple_quadratic_bo_steps`?

File: tests/integration/test_ask_tell_optimization.py [hunk 5]
```diff
@@ -113,6 +143,11 @@ def test_ask_tell_optimization_finds_minima_of_the_scaled_branin_function(
 
     search_space = BRANIN_SEARCH_SPACE
 
+    # for speed reasons we sometimes test with a simple quadratic defined on the
+    # same search space as branin
+    def simple_quadratic(x: TensorType) -> TensorType:
+        return -tf.math.reduce_sum(x, axis=-1, keepdims=True) ** 2
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]: Maybe move it to test/utils package?

File: tests/integration/test_ask_tell_optimization.py [hunk 6]
```diff
@@ -97,7 +95,39 @@
...
+    | Callable[
+        [],
+        AcquisitionRule[State[TensorType, AsynchronousRuleState | TrustRegion.State], Box],
+    ],
+) -> None:
+    _test_ask_tell_optimization_finds(False, min(num_steps, 5), reload_state, acquisition_rule_fn)
+
+
+def _test_ask_tell_optimization_finds(
+    branin: bool,
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]:
```suggestion
    optimize_branin: bool,
```

File: tests/integration/test_ask_tell_optimization.py [hunk 7]
```diff
@@ -164,8 +199,13 @@ def build_model(data: Dataset) -> GaussianProcessRegression:
...
+    if branin:
+        relative_minimizer_err = tf.abs((best_x - BRANIN_MINIMIZERS) / BRANIN_MINIMIZERS)
+        # these accuracies are the current best for the given number of optimization steps,
+        # which makes this is a regression test
+        assert tf.reduce_any(tf.reduce_all(relative_minimizer_err < 0.05, axis=-1), axis=0)
+        npt.assert_allclose(best_y, SCALED_BRANIN_MINIMUM, rtol=0.005)
+    else:
+        # currently assume that every rule should be able to solve this in 5 steps
+        npt.assert_allclose(best_x, tf.constant([1.0, 1.0], tf.float64), rtol=0.05)
+        npt.assert_allclose(best_y, tf.constant([-4.0], tf.float64), rtol=0.05)
```
Comment (apaleyes) [inline tests/integration/test_ask_tell_optimization.py]: Again, would be nice to put these to constants and move along the function definition. In fact, maybe it's worth having this funciton just as any other objective in `single_objectives`?

**Review by apaleyes — COMMENTED** [review]
Looks great! Below are just some ideas about location and naming of things

File: tests/integration/test_ask_tell_optimization.py (hunk 6 above)
Comment (apaleyes) [reply tests/integration/test_ask_tell_optimization.py]: another option: `is_branin`

**Review by apaleyes — APPROVED** [review]

**Pull request merged**